packages/vitest/src/node/cli.ts

```diff
@@ -33,6 +33,9 @@ cli
   .option('--environment <env>', 'runner environment (default: node)')
   .option('--passWithNoTests', 'pass when no tests found')
   .option('--allowOnly', 'Allow tests and suites that are marked as only (default: !process.env.CI)')
+  .option('--onlyChanged', 'Run tests that are affected by the changed files (default: false)')
+  .option('--lastCommit', 'Run tests that are affected by files changed in the last commit (default: false)')
+  .option('--changedSince <commit>', 'Run tests that are affected by files changed since commit hash or branch')
```

Comment:
This feels a bit duplicated to me, do you think we could have a single command like `--changedSince [commit]` with a convention like `HEAD^3` or even simpler like `-3` for the previous 3 commits?

Reply:
I would expect people to use actual hashes? I think testing by 'previous # commits' is not really an issue? You will run this only in CI, and CI usually have some global variables that contain hashes you can use, like `GITHUB_SHA`

What exactly do you find repetitive?

Reply:
I think `--changedSince` is the only one that really should be run in CI

`--onlyChanged` will run tests against files that are just that - changed, not commited or anything
`--lastCommit` is also to run it maybe on hook or if you forgot to test with `--onlyChanged`

Reply:
I am still not sure if to have 3 options to do the same thing. It's also not clear to me if and how I should use them together.

I am thinking it like:

- `--changed` - from the latest commit, for unstaged changes. If the stage is clear, for the previous commit.
- `--changed ^3` - changes in the previous 3 commits (convention could be discussed)
- `--changed 5963a39` - since SHA
- `--changed main` - changes against main branch (or tags)

Reply:
I don't see a profit in cramming different usecases into a single option. Wouldn't it increase code heaviness? Also your suggestions solution makes a lot of assumptions, and I need a book to understand it (like running `--changed` without options)

I think having 3 different options is more approachable for users (I disagree they do the same thing). Especially coming from jest. Also I don't want to fix what is not broken (like previous 3 commits - who need this? what its usecase?)

For me you solution looks like instead of having 3 different functions with distinct names use one - I really don't see how this is better?

Reply:
> It's also not clear to me if and how I should use them together.

Well, you wouldn't. `--onlyChanged` can be used with both, it will be overriden.

But we also have `--watch` and `--run` options that you can't use at the same time.

Reply:
What's behaviors would it be with?

- `vitest --lastCommit --changedSince fff`
- `vitest --lastCommit`
- `vitest --onlyChanged --lastCommit --changedSince fff`

3 options you got 8 different combinations (that's a multiplier if you have other options).

> I need a book to understand it (like running --changed without options)

It's common to have optional arguments as we had for `--api [api]`. I don't understand what is confusing.

Reply:
> It's common to have optional arguments as we had for `--api [api]`. I don't understand what is confusing.

It's not options that are confusing. Described behaviour is - why would it behave differently depending on staged changes?

> What's behaviors would it be with?

What's important (specific) will override what's less important (specific)

Reply:
Ok, I researched a little bit more, and what about:

- `--changed` - runs tests against unstaged changes
- `--changed hash/branch/HEAD~#` - runs against hash/branch (command is the same)

Documentation can describe how to do common tasks (like from last, use `HEAD`, `HEAD~1`)

Reply:
Yeah, that's what I was proposing.